### PR TITLE
Correctly use the rotation vectors

### PR DESCRIPTION
**Pull request by trickeydan**
It should be noted whilst this is being developed that `yaw_pitch_roll` is in the order `z`,`y`,`x` as this is standard in vision applications. Search for "Intrinsic Tait-Bryan Angles" for more information.

### CODE REVIEW

**Comment by RealOrangeOne** [comment]
This amount of duplication is fine IMO

**Comment by trickeydan** [comment]
(I gitted bad, please ignore the closing /reopening)

File: zoloto/coords.py [hunk 1]
```diff
@@ -1,13 +1,84 @@
...
-
 ThreeDCoordinates = spaced_coordinate("ThreeDCoordinates", "xyz")
 
 Spherical = NamedTuple("Spherical", [("rot_x", float), ("rot_y", float), ("dist", int)])
+
+
+class Orientation:
+    """The orientation of an object in 3-D space."""
+
+    def __init__(self, *rvec) -> None:
```
Comment (RealOrangeOne) [inline zoloto/coords.py]: is `rvec` always `rot_x, rot_y, rot_z`? If so would be nice to constrain the arguments (and therefore the types)

File: zoloto/coords.py [hunk 2]
```diff
@@ -1,13 +1,84 @@
...
+            pitch:  rotation angle around the y'-axis in radians, in the range `[-pi/2, -pi/2]`
+            roll:   rotation angle around the x''-axis in radians, in the range `[-pi, pi]`
+        """
+        return self._quaternion.yaw_pitch_roll
+
+    def __iter__(self) -> Iterator[float]:
+        return iter([self.rot_x, self.rot_y, self.rot_z])
+
+    @cached_property
+    def rotation_matrix(self) -> ndarray:
```
Comment (RealOrangeOne) [inline zoloto/coords.py]: The rotation matrix should always be a 3-element tuple, right? Would be nice to cast it to that, or just expose the originally passed matrix?

File: zoloto/coords.py [hunk 3]
```diff
@@ -1,13 +1,84 @@
...
+class Orientation:
+    """The orientation of an object in 3-D space."""
+
+    def __init__(self, *rvec) -> None:
+        rotation_matrix, _ = Rodrigues(rvec)
+        self._quaternion = Quaternion(matrix=rotation_matrix)
+
+    @property
+    def rot_x(self) -> float:
+        """Get rot_x aka. roll."""
```
Comment (RealOrangeOne) [inline zoloto/coords.py]: I assume all these values are in radians? Would be nice to note that in the docstring

File: setup.py [hunk 4]
```diff
@@ -18,6 +18,7 @@
         "cached-property>=1.5",
         "coordinates>=0.3.0",
         "fastcache>=1.1",
+        "pyquaternion>0.9.2",
```
Comment (RealOrangeOne) [inline setup.py]: Typo?

```suggestion
        "pyquaternion>=0.9.2",
```

**Comment by RealOrangeOne** [comment]
Couple comments on the code. I'll load up a preview later to check the values out look sane

**Comment by trickeydan** [comment]
> Couple comments on the code. I'll load up a preview later to check the values out look sane

You might find [this](https://gist.github.com/trickeydan/aeb67c406a0e46007b618528a59e354a) useful when testing. It's the program I was using to debug things.

File: zoloto/coords.py (hunk 1 above)
Comment (trickeydan) [reply zoloto/coords.py]: There was a reason I didn't do this.

Technically they aren't rot_x, rot_y, rot_z in the `rvec`, but I will find the correct terminology and give this another go.

**Comment by RealOrangeOne** [comment]
I've given this a quick runthrough locally, and it seems to be working ok. But then again I thought the previous values were correct. The values don't look entirely crazy, which is something.

From my _brief_ understanding, this is correct, as matrix != angles.

It's very difficult to tell whether the snapshot results are right. In think in future we should collect some more accurate test images to fully prove this.

1 thing which does weird me out is that a marker facing the camera has a `rot_x` of -180/180 degrees, which feels strange. Should this be 0? Might just be bad markers on my end.

**Comment by trickeydan** [comment]
> 1 thing which does weird me out is that a marker facing the camera has a `rot_x` of -180/180 degrees, which feels strange. Should this be 0? Might just be bad markers on my end.

I believe that is correct in the system that we are using. 

The system of rotx,y, z is generally known as Euler angles, however there are many different conventions for these, the only standard of which are "intrinsic tait Bryan angles". These are what we are using. 

I will double check when make the requested changes.

**Comment by trickeydan** [comment]
> I will double check when make the requested changes.

I have double checked, and this is actually what we would expect. It will be +/-180deg.

Explanation:

image

In the image of the plane, imagine that you are staring directly at the propeller on the nose of the plane. The marker is on the nose of the plane.

- The x axis extends to your left and right.
- The y axis extends above and below you.
- The z axis goes in front and behind you.

The value of `rot_a` represents the "tilt" of the plane around the "a" axis.

File: zoloto/coords.py (hunk 1 above)
Comment (trickeydan) [reply zoloto/coords.py]: 0478026

File: zoloto/coords.py (hunk 3 above)
Comment (trickeydan) [reply zoloto/coords.py]: 3373076

**Comment by RealOrangeOne** [comment]
> The value of rot_a represents the "tilt" of the plane around the "a" axis.

This I already knew, but I guess it's more about it's the rotation relative to the object, so in this case the marker is facing towards us, so it's already had a 180 degree rotation around said axis. 

Makes sense, nice writeup!

File: zoloto/coords.py [hunk 5]
```diff
@@ -1,13 +1,107 @@
...
-Orientation = NamedTuple(
-    "Orientation", [("rot_x", float), ("rot_y", float), ("rot_z", float)]
-)
-
 ThreeDCoordinates = spaced_coordinate("ThreeDCoordinates", "xyz")
 
 Spherical = NamedTuple("Spherical", [("rot_x", float), ("rot_y", float), ("dist", int)])
+
+ThreeTuple = Tuple[float, float, float]
+RotationMatrix = Tuple[ThreeTuple, ThreeTuple, ThreeTuple]
```
Comment (RealOrangeOne) [inline zoloto/coords.py]: Would these be better defined as `NamedTuple`?

Comment (trickeydan) [reply zoloto/coords.py]: Unlikely, its a rotation matrix, so values have no other name than their position in the matrix.

**Pull request merged**